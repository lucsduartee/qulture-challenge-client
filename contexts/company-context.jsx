import React, { createContext, useReducer } from 'react'

export const CompaniesContext = createContext()
export const CompaniesDispatchContext = createContext()

const initialCompanies = {
  selectedCompany: null,
  companies: [],
};

function companiesReducer(companies, action) {
  switch (action.type) {
    case 'charged': {
      return {
        ...companies,
        companies: action.data.companies
      }
    }
    case 'created': {
      return {
        ...companies,
        companies: [...companies.companies, action.data.company]
      }
    }
    case 'selected': {
      return {
        ...companies,
        selectedCompany: action.data.company
      }
    }
    default: {
      throw Error('Unknown action: ' + action.type);
    }
  }
}

export default function CompaniesProvider({ children }) {
  const [companies, dispatch] = useReducer(
    companiesReducer,
    initialCompanies,
  );

  return (
    <CompaniesContext.Provider value={companies}>
      <CompaniesDispatchContext.Provider value={dispatch}>
        { children }
      </CompaniesDispatchContext.Provider>
    </CompaniesContext.Provider>
  )
}