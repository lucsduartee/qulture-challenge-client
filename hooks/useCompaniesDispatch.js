import { useContext } from 'react'
import { CompaniesDispatchContext } from '@/contexts/company-context';

const useCompaniesDispatch = () => {
  const dispatch = useContext(CompaniesDispatchContext);
  if (!dispatch) {
    throw new Error('CompaniesDispatchContext must be used within a CompaniesProvider');
  }
  return dispatch;
}

export default useCompaniesDispatch;
