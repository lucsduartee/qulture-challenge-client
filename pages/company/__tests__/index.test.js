import * as React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import Company from '../index';
import { CompaniesContext } from '@/contexts/company-context';
import useCompaniesDispatch from '@/hooks/useCompaniesDispatch';

jest.mock('../../../hooks/useCompaniesDispatch');

jest.mock('next/router', () => jest.requireActual('next-router-mock'))

global.fetch = jest.fn(() =>
  Promise.resolve({
    ok: true,
    json: () => Promise.resolve([{ id: 1, name: 'Company A' }]),
  })
);

const mockCompanies = [
  { id: 1, name: 'Company A', total_employees: 10, leaders_count: 2, followers_count: 8 },
  { id: 2, name: 'Company B', total_employees: 20, leaders_count: 3, followers_count: 17 },
];

const mockSelectedCompany = mockCompanies[0];

const mockDispatch = jest.fn();

const renderCompanyComponent = (companies = [], selectedCompany = null) => {
  return render(
    <CompaniesContext.Provider value={{ companies, selectedCompany }}>
      <Company />
    </CompaniesContext.Provider>
  );
};

describe('Company Component', () => {
  beforeEach(() => {
    useCompaniesDispatch.mockReturnValue(mockDispatch);
    fetch.mockClear();
    mockDispatch.mockClear();
  });

  it('matches snapshot', () => {
    const tree = renderCompanyComponent(mockCompanies, mockSelectedCompany);

    expect(tree).toMatchSnapshot();
  });
  

  it('renders the component correctly', () => {
    renderCompanyComponent(mockCompanies, mockSelectedCompany);

    expect(screen.getByText('Cadastrar Empresa')).toBeInTheDocument();
    expect(screen.getByLabelText('Nome da Empresa')).toBeInTheDocument();
    expect(screen.getByText('Cadastrar')).toBeInTheDocument();
    expect(screen.getByText('Lista de Empresas')).toBeInTheDocument();
    expect(screen.getByText('Detalhes da Empresa')).toBeInTheDocument();
  });

  it('fetches companies on mount', async () => {
    renderCompanyComponent();

    await waitFor(() => {
      expect(fetch).toHaveBeenCalledWith(`${process.env.NEXT_PUBLIC_QULTURE_API_HOST}/api/companies`, {
        method: 'GET',
      });
      expect(mockDispatch).toHaveBeenCalledWith({
        type: 'charged',
        data: {
          companies: [{ id: 1, name: 'Company A' }],
        },
      });
    });
  });

  it('submits a new company', async () => {
    renderCompanyComponent();

    const input = screen.getByLabelText('Nome da Empresa');
    const button = screen.getByText('Cadastrar');

    fireEvent.change(input, { target: { value: 'New Company' } });
    fireEvent.click(button);

    await waitFor(() => {
      expect(fetch).toHaveBeenCalledWith(`${process.env.NEXT_PUBLIC_QULTURE_API_HOST}/api/companies`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ name: 'New Company' }),
      });
      expect(mockDispatch).toHaveBeenCalledWith({
        type: 'created',
        data: {
          company: [{ id: 1, name: 'Company A' }],
        },
      });
    });
  });

  it('selects a company from the list', () => {
    renderCompanyComponent(mockCompanies);

    const companyItem = screen.getByText('Company A');
    fireEvent.click(companyItem);

    expect(mockDispatch).toHaveBeenCalledWith({
      type: 'selected',
      data: {
        company: mockCompanies[0],
      },
    });
  });
});
