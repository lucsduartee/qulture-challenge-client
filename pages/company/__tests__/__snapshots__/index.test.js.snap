// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Company Component matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root css-jumd57"
      >
        <div
          class="MuiBox-root css-0"
        >
          <h5
            class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-1f9xtph-MuiTypography-root"
          >
            Cadastrar Empresa
          </h5>
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-dzmwfx-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for=":r0:"
              id=":r0:-label"
            >
              Nome da Empresa
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1blp12k-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r0:"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-81qg8w"
                >
                  <span>
                    Nome da Empresa
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-5kl8be-MuiButtonBase-root-MuiButton-root"
            id=":r1:"
            tabindex="0"
            type="button"
          >
            Cadastrar
          </button>
        </div>
        <hr
          class="MuiDivider-root MuiDivider-fullWidth css-1jpc804-MuiDivider-root"
        />
        <div
          class="MuiBox-root css-0"
        >
          <h5
            class="MuiTypography-root MuiTypography-h5 css-16xl4zq-MuiTypography-root"
          >
            Lista de Empresas
          </h5>
          <ul
            class="MuiList-root MuiList-padding css-b7dj7n-MuiList-root"
          >
            <li
              class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-100imhj-MuiListItem-root"
            >
              Company A
            </li>
            <li
              class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-100imhj-MuiListItem-root"
            >
              Company B
            </li>
          </ul>
        </div>
        <hr
          class="MuiDivider-root MuiDivider-fullWidth css-1jpc804-MuiDivider-root"
        />
        <div
          class="MuiBox-root css-0"
        >
          <h5
            class="MuiTypography-root MuiTypography-h5 css-16xl4zq-MuiTypography-root"
          >
            Detalhes da Empresa
          </h5>
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1j74uew-MuiPaper-root-MuiCard-root"
            style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
          >
            <div
              class="MuiCardContent-root css-1lt5qva-MuiCardContent-root"
            >
              <h6
                class="MuiTypography-root MuiTypography-h6 css-1miy0lu-MuiTypography-root"
              >
                <b>
                  Nome
                </b>
              </h6>
              <p
                class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
              >
                Company A
              </p>
              <hr
                class="MuiDivider-root MuiDivider-fullWidth css-zzw4lh-MuiDivider-root"
              />
              <p
                class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
              >
                <b>
                  Total de Colaboradores:
                </b>
                 
                10
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
              >
                <b>
                  LÃ­deres:
                </b>
                 
                2
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
              >
                <b>
                  Liderados:
                </b>
                 
                8
              </p>
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-s1gs1n-MuiButtonBase-root-MuiButton-root"
                id=":r2:"
                tabindex="0"
                type="button"
              >
                Ver mais detalhes
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-jumd57"
    >
      <div
        class="MuiBox-root css-0"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-1f9xtph-MuiTypography-root"
        >
          Cadastrar Empresa
        </h5>
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-dzmwfx-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":r0:"
            id=":r0:-label"
          >
            Nome da Empresa
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1blp12k-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r0:"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-81qg8w"
              >
                <span>
                  Nome da Empresa
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-5kl8be-MuiButtonBase-root-MuiButton-root"
          id=":r1:"
          tabindex="0"
          type="button"
        >
          Cadastrar
        </button>
      </div>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-1jpc804-MuiDivider-root"
      />
      <div
        class="MuiBox-root css-0"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5 css-16xl4zq-MuiTypography-root"
        >
          Lista de Empresas
        </h5>
        <ul
          class="MuiList-root MuiList-padding css-b7dj7n-MuiList-root"
        >
          <li
            class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-100imhj-MuiListItem-root"
          >
            Company A
          </li>
          <li
            class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-100imhj-MuiListItem-root"
          >
            Company B
          </li>
        </ul>
      </div>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-1jpc804-MuiDivider-root"
      />
      <div
        class="MuiBox-root css-0"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5 css-16xl4zq-MuiTypography-root"
        >
          Detalhes da Empresa
        </h5>
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1j74uew-MuiPaper-root-MuiCard-root"
          style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
        >
          <div
            class="MuiCardContent-root css-1lt5qva-MuiCardContent-root"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 css-1miy0lu-MuiTypography-root"
            >
              <b>
                Nome
              </b>
            </h6>
            <p
              class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
            >
              Company A
            </p>
            <hr
              class="MuiDivider-root MuiDivider-fullWidth css-zzw4lh-MuiDivider-root"
            />
            <p
              class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
            >
              <b>
                Total de Colaboradores:
              </b>
               
              10
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
            >
              <b>
                LÃ­deres:
              </b>
               
              2
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
            >
              <b>
                Liderados:
              </b>
               
              8
            </p>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-s1gs1n-MuiButtonBase-root-MuiButton-root"
              id=":r2:"
              tabindex="0"
              type="button"
            >
              Ver mais detalhes
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
